packages/toolkit/src/query/core/buildThunks.ts

```diff
@@ -500,20 +500,30 @@ In the case of an unhandled error, no tags will be "provided" or "invalidated".`
   }
 }
 
+type CalculatableAction = UnwrapPromise<
+  ReturnType<ReturnType<QueryThunk>> | ReturnType<ReturnType<MutationThunk>>
+>
 export function calculateProvidedByThunk(
-  action: UnwrapPromise<
-    ReturnType<ReturnType<QueryThunk>> | ReturnType<ReturnType<MutationThunk>>
-  >,
+  action: CalculatableAction,
   type: 'providesTags' | 'invalidatesTags',
   endpointDefinitions: EndpointDefinitions,
   assertTagType: AssertTagTypes
 ) {
+  const isQuery = (action: CalculatableAction) =>
+    action.meta.arg.type === 'query'
+  const isFulfilledQuery = (action: any): action is QueryThunk =>
+    isQuery(action)
+  const isFulfilledMutation = (action: any): action is MutationThunk =>
+    !isQuery(action)
+
   return calculateProvidedBy(
     endpointDefinitions[action.meta.arg.endpointName][type],
     isFulfilled(action) ? action.payload : undefined,
     isRejectedWithValue(action) ? action.payload : undefined,
     action.meta.arg.originalArgs,
-    action.meta,
+    isAnyOf(isRejected, isFulfilledQuery, isFulfilledMutation)(action)
+      ? action.meta.baseQueryMeta
+      : undefined,
```

Comment:
I don't know if all of this ceremony is really necessary, but it's probably better than `(action.meta as any).baseQueryMeta` or something like that. Very open to ideas here @phryneas if you have one :)

Reply:
I think that can be done with less changes - I pushed my suggestion.

Reply:
Thanks! I need to get a 'ways to narrow types' checklist solidified in my brain.